staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go

```diff
@@ -254,13 +256,24 @@ func ResetObjectMetaForStatus(meta, existingMeta Object) {
 }
 
 // MarshalJSON implements json.Marshaler
+// MarshalJSON may get called on pointers or values, so implement MarshalJSON on value.
+// http://stackoverflow.com/questions/21390979/custom-marshaljson-never-gets-called-in-go
 func (f Fields) MarshalJSON() ([]byte, error) {
-	return json.Marshal(&f.Map)
+	if f.Raw == nil {
+		return []byte("null"), nil
+	}
+	return f.Raw, nil
 }
 
 // UnmarshalJSON implements json.Unmarshaler
 func (f *Fields) UnmarshalJSON(b []byte) error {
-	return json.Unmarshal(b, &f.Map)
+	if f == nil {
+		return errors.New("metav1.Fields: UnmarshalJSON on nil pointer")
+	}
+	if !bytes.Equal(b, []byte("null")) {
+		f.Raw = append(f.Raw[0:0], b...)
```

Comment:
Does this combo loose the unset / null distinction? I think it might? Will our get/put path treat an explicit null as an attempt to delete the field?

Reply:
That's a good question. I copied this code from the rawExtension marshaller. I'm not sure if our get/put path needs to worry about this field being unset or null though, since it is already nested within managedFields